### ISSUE TITLE
[go][channel] - parallel as well as memory bounded

### ISSUE THREAD

**Issue by AmitKumarDas** [AmitKumarDas/fun-with-programming#1]
refer - https://blog.golang.org/pipelines

```go
// producer
```

```go
type result struct {
    path string
    sum  [md5.Size]byte
    err  error
}
```

```go
func sumFiles(done <-chan struct{}, root string) (<-chan result, <-chan error) {
    c := make(chan result)  // unbuffered
    errc := make(chan error, 1) // buffered

    // with above channels in place
    // start a goroutine & MD5 files in parallel
    go func() {
        var wg sync.WaitGroup

        // func within a func
        // see nested use of wg
        err := filepath.Walk(root, func(path string, i os.FileInfo, err error) error {
            if err != nil {
                return err
            }
            if !i.Mode().IsRegular() {
                return nil
            }

            // each file is processed in a goroutine
            // what if more memory than what is available [f5]
            wg.Add(1)
            go func() {
                data, err := ioutil.ReadFile(path)
                select {
                case c <- result{path, md5.Sum(data), err}:
                case <-done:
                }
                wg.Done()
            }()
            // Abort the walk if done is closed.
            select {
            case <-done:
                return errors.New("walk canceled") // not an err to worry [??]
            default:
                return nil
            }
        })

        // Walk has returned, so all calls to wg.Add are done.  Start a
        // goroutine to close c once all the sends are done.
        go func() {
            wg.Wait() // blocking
            close(c)
        }()

        // No select needed here, since errc is buffered.
        errc <- err
    }()

    return c, errc   // c is closeable
}
```

**Comment by AmitKumarDas** [comment]
```go
// consumer
```

```go
func MD5All(root string) (map[string][md5.Size]byte, error) {
    // MD5All closes the done channel when it returns; it may do so before
    // receiving all the values from c and errc.
    done := make(chan struct{})
    defer close(done)

    c, errc := sumFiles(done, root)

    m := make(map[string][md5.Size]byte)
    for r := range c { // loop till c closes
        if r.err != nil {
            return nil, r.err
        }
        m[r.path] = r.sum
    }
    if err := <-errc; err != nil { // non blocking since errc is buffered
        return nil, err
    }
    return m, nil
}
```

**Comment by AmitKumarDas** [comment]
```go
// bounded alternative
// keep goroutines under check
```

```go
func walkFiles(done <-chan struct{}, root string) (<-chan string, <-chan error) {
    paths := make(chan string) // unbuffered
    errc := make(chan error, 1) // buffered

    go func() { // goroutine at root
        // Close the paths channel after Walk returns.
        defer close(paths)

        // select not needed for this send
        // since errc is buffered
        errc <- filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
            if err != nil {
                return err
            }
            if !info.Mode().IsRegular() {
                return nil
            }
            select { // blocking select block
            case paths <- path:   // unbuffered
            case <-done:  // unbuffered
                return errors.New("walk canceled")
            }
            return nil
        })
    }()
    return paths, errc
}
```

```go
func MD5All(root string) (map[string][md5.Size]byte, error) {
    // Start a fixed number of goroutines to read and digest files.
    c := make(chan result)
    var wg sync.WaitGroup
    const numDigesters = 20
    wg.Add(numDigesters)
    for i := 0; i < numDigesters; i++ {
        go func() {
            digester(done, paths, c)
            wg.Done()
        }()
    }
    go func() {
        wg.Wait()
        close(c)
    }()
    
    m := make(map[string][md5.Size]byte)
    for r := range c {
        if r.err != nil {
            return nil, r.err
        }
        m[r.path] = r.sum
    }
    // Check whether the Walk failed.
    if err := <-errc; err != nil {
        return nil, err
    }
    return m, nil
}
```

```go
func digester(done <-chan struct{}, paths <-chan string, c chan<- result) {
    for path := range paths {
        data, err := ioutil.ReadFile(path)
        select {
        case c <- result{path, md5.Sum(data), err}:
        case <-done:
            return
        }
    }
}
```
